smart-contracts/contracts/Unlock.sol

```diff
@@ -126,7 +148,55 @@ contract Unlock is
     initializer()
   {
     // We must manually initialize Ownable
-    Ownable.__initializeOwnable(_unlockOwner);
+    UnlockOwnable.__initializeOwnable(_unlockOwner);
+  }
+
+  function initializeProxyAdmin() public {
+    require(proxyAdminAddress == address(0), "ProxyAdmin already deployed");
+    _deployProxyAdmin();
+  }
+
+  /**
+  * @dev Deploy the ProxyAdmin contract that will manage lock templates upgrades
+  * This deploys an instance of ProxyAdmin used by PublicLock transparent proxies.
+  */
+  function _deployProxyAdmin() private returns(address) {
+    proxyAdmin = new ProxyAdmin();
+    proxyAdminAddress = address(proxyAdmin);
+    return address(proxyAdmin);
+  }
+
+  /**
+  * @dev Helper to get the version number of a template from his address
+  */
+  function publicLockVersions(address _impl) external view returns(uint16) {
+    return _publicLockVersions[_impl];
+  }
+
+  /**
+  * @dev Helper to get the address of a template based on its version number
+  */
+  function publicLockImpls(uint16 _version) external view returns(address) {
+    return _publicLockImpls[_version];
+  }
+
+  /**
+  * @dev Registers a new PublicLock template immplementation
+  * The template is identified by a version number
+  * Once registered, the template can be used to upgrade an existing Lock
+  */
+  function addLockTemplate(address impl, uint16 version) public onlyOwner {
+    require(impl != address(0), "impl address can not be 0x");
+    require(version != 0, "version can not be 0");
```

Comment:
I wonder how "useful" that is?
Maybe we should check if this is indeed a smart contract address and if this is indeed a `PublickLock`? but it's probably hard to do?

Reply:
there is no proper way I know of in Solidity to check if an address actually exists or if it is a contract. You can check but that will open up to some possible attacks (by faking a contract).

Reply:
regarding the usefulness, it is more of my own coding convention to check the validity of a function params. That may come from too much Javascript lately :P Agree that as Solidity is strongly typed that may not be so useful

Reply:
ok I think since this is `onlyOwner` we can "trust" that they do good things? (and allow them to overwrite below always). So worst case scenario: they waste gas?

Reply:
So basically I would remove the checks

Reply:
yep fine with me